kornia/augmentation/_2d/base.py

```diff
@@ -21,27 +26,104 @@ class AugmentationBase2D(_AugmentationBase):
         """Convert any incoming (H, W), (C, H, W) and (B, C, H, W) into (B, C, H, W)."""
         _validate_input_dtype(input, accepted_dtypes=[float16, float32, float64])
         return _transform_input(input)
 
+
+class RigidAffineAugmentationBase2D(AugmentationBase2D):
+    r"""AugmentationBase2D base class for rigid/affine augmentation implementations.
+
+    RigidAffineAugmentationBase2D enables routined transformation with given transformation matrices
+    for different data types like masks, boxes, and keypoints.
+
+    Args:
+        p: probability for applying an augmentation. This param controls the augmentation probabilities
+          element-wise for a batch.
+        p_batch: probability for applying an augmentation to a batch. This param controls the augmentation
+          probabilities batch-wise.
+        same_on_batch: apply the same transformation across the batch.
+        keepdim: whether to keep the output shape the same as input ``True`` or broadcast it to the batch
+          form ``False``.
+    """
+
+    _transform_matrix: Optional[Tensor]
+
+    @property
+    def transform_matrix(self) -> Optional[Tensor]:
+        return self._transform_matrix
+
     def identity_matrix(self, input) -> Tensor:
         """Return 3x3 identity matrix."""
-        return kornia.eye_like(3, input)
+        return eye_like(3, input)
+
+    def compute_transformation(self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any]) -> Tensor:
+        raise NotImplementedError
+
+    def generate_transformation_matrix(self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any]) -> Tensor:
+        """Generate transformation matrices with the given input and param settings."""
+        # Note about the explicit type conversions here:
+        # If this code is run in an autocast-enabled region, the transformation matrix or the output
+        # may be float16 even though the input was float32 (e.g. torch.mm produces float16 output),
+        # see also the documentation on autocasting: https://pytorch.org/docs/stable/amp.html.
+        # It may be unexpected for the user if the output type changes and it can also lead to errors
+        # for the index_put operation below (see also https://github.com/kornia/kornia/issues/1737)
+        # In case the type already matches the input type, the conversions are no-ops
+
+        to_apply = params['batch_prob']
+        in_tensor = self.transform_tensor(input)
+        if not to_apply.any():
+            trans_matrix = self.identity_matrix(in_tensor)
+        elif to_apply.all():
+            trans_matrix = self.compute_transformation(in_tensor, params=params, flags=flags).type(in_tensor.dtype)
+        else:
+            trans_matrix = self.identity_matrix(in_tensor)
+            trans_matrix = trans_matrix.index_put(
+                (to_apply,),
+                self.compute_transformation(in_tensor[to_apply], params=params, flags=flags).type(in_tensor.dtype),
```

Comment:
not related with autocast, but I think this is not working  -- this seems to fix some 2d geometric cases, but I don't know if it's the right thing to do (at a first glance seems to be better to do this inside each `compute_transformation`)
```suggestion
                self.compute_transformation(in_tensor[to_apply], params=params, flags=flags)[to_apply],
```

without this occurs shape mismatch between `trans_matrix[to_apply]` and `self.compute_transformation(...)` -- tested on affine and perspective

Reply:
It should be working as it is from the original code:

https://github.com/kornia/kornia/blob/f794dbd25ea3ce8439c5e3b0a1b86405edd8fb17/kornia/augmentation/base.py#L268-L271

Also, if `in_tensor` is `8CHW`, and `to_apply` is `[T,T,T,T,F,F,F,F]`, `in_tensor[to_apply]` will be `4CHW`, how can it be applied again with the `to_apply`?

Reply:
I probably just set up wrong the test, I just copied the autocast one from the base  -- https://github.com/shijianjian/kornia/compare/feat/aug_base_refactor...johnnv1:kornia:fix-autocast